builtin/aws/lambda/platform.go

```diff
@@ -0,0 +1,706 @@
+	}
+
+	_, err = svc.AttachRolePolicy(aInput)
+	if err != nil {
+		return "", err
+	}
+
+	L.Info("attached execution role policy")
+
+	return roleArn, nil
+}
+
+// Deploy deploys an image to AWS Lambda.
+func (p *Platform) Deploy(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	img *ecr.Image,
+	deployConfig *component.DeploymentConfig,
+	ui terminal.UI,
+) (*Deployment, error) {
+	// Start building our deployment since we use this information
+	deployment := &Deployment{}
+	id, err := component.Id()
+	if err != nil {
+		return nil, err
+	}
+
+	sess, err := utils.GetSession(&utils.SessionConfig{
+		Region: p.config.Region,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	roleArn := p.config.RoleArn
+
+	if roleArn == "" {
+		arn, err := p.SetupRole(log, src, sess)
+		if err != nil {
+			return nil, err
+		}
+
+		roleArn = arn
+	}
+
+	mem := int64(p.config.Memory)
+	if mem == 0 {
+		mem = DefaultMemory
+	}
+
+	timeout := int64(p.config.Timeout)
+	if timeout == 0 {
+		timeout = DefaultTimeout
+	}
+
+	lamSvc := lambda.New(sess)
+	curFunc, err := lamSvc.GetFunction(&lambda.GetFunctionInput{
+		FunctionName: aws.String(src.App),
+	})
+
+	var funcarn string
+
+	// If the function exists, we update it's code rather than create a new one.
+	// Will then publish this new code to create a stable identifier for it.
+	if err == nil {
+
+		var reset bool
+		var update lambda.UpdateFunctionConfigurationInput
+
+		if *curFunc.Configuration.MemorySize != mem {
+			update.MemorySize = aws.Int64(mem)
+			reset = true
+		}
+
+		if *curFunc.Configuration.Timeout != timeout {
+			update.Timeout = aws.Int64(timeout)
+			reset = true
+		}
+
+		if reset {
+			update.FunctionName = curFunc.Configuration.FunctionArn
+
+			_, err = lamSvc.UpdateFunctionConfiguration(&update)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to update function configuration")
+			}
+		}
+
+		funcCfg, err := lamSvc.UpdateFunctionCode(&lambda.UpdateFunctionCodeInput{
+			FunctionName: aws.String(src.App),
+			ImageUri:     aws.String(img.Name()),
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		funcarn = *funcCfg.FunctionArn
+
+		if err != nil {
+			return nil, err
+		}
+
+		ui.Output("Updated Lambda Function: %s", funcarn)
+	} else {
+		funcOut, err := lamSvc.CreateFunction(&lambda.CreateFunctionInput{
+			Description:  aws.String(fmt.Sprintf("waypoint %s", src.App)),
+			FunctionName: aws.String(src.App),
+			Role:         aws.String(roleArn),
+			Timeout:      aws.Int64(timeout),
+			MemorySize:   aws.Int64(mem),
+			Tags: map[string]*string{
+				"waypoint.app": aws.String(src.App),
+			},
+			PackageType: aws.String("Image"),
+			Code: &lambda.FunctionCode{
+				ImageUri: aws.String(img.Name()),
+			},
+			ImageConfig: &lambda.ImageConfig{},
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		funcarn = *funcOut.FunctionArn
+
+		ui.Output("Created Lambda Function: %s", funcarn)
+	}
+
+	st := ui.Status()
```

Comment:
If possible, I'd avoid Status and use `StepGroup` throughout this thing instead.